pkg/util/push/push.go

```diff
@@ -85,43 +84,38 @@ func handler(maxRecvMsgSize int,
 				logger = log.WithSourceIPs(source, logger)
 			}
 		}
-		bufHolder := bufferPool.Get().(*bufHolder)
-		var req mimirpb.PreallocWriteRequest
-		buf, err := parser(ctx, r, maxRecvMsgSize, bufHolder.buf, &req)
-		if err != nil {
-			level.Error(logger).Log("err", err.Error())
-
-			// Check for httpgrpc error.
-			if resp, ok := httpgrpc.HTTPResponseFromError(err); ok {
-				http.Error(w, string(resp.Body), int(resp.Code))
-			} else {
-				http.Error(w, err.Error(), http.StatusBadRequest)
+		supplier := func() (*mimirpb.WriteRequest, func(), error) {
+			bufHolder := bufferPool.Get().(*bufHolder)
+			var req mimirpb.PreallocWriteRequest
+			buf, err := parser(ctx, r, maxRecvMsgSize, bufHolder.buf, &req)
+			if err != nil {
+				// Check for httpgrpc error, default to client error if parsing failed
```

Comment:
This looks like a logic change to me. Before we were guaranteeing to always return a HTTP error, while now we could return a httpgrpc error. Isn't a regression?

Reply:
that's true, I had conflated the httpgrpc and http packages. It also means we also need a test for this, no?

Reply:
part of the translation was moved elsewhere actually 

https://github.com/grafana/mimir/blob/bd40d46d9c8991a2fef7f529f35674eb6db33127/pkg/util/push/push.go#L119-L132

so what you're commenting on is the error that will be returned to whoever did `_, err := push.Request.WriteRequest()`. This error wasn't previously exposed outside the push package, so it's new behaviour.

I added more tests in 70d8de261d7b7faa0d963043b8271f1cb78f95f3

Reply:
You're right. LGTM, thanks.